File: 3720-lexicographically-smallest-permutation-greater-than-target/3720-lexicographically-smallest-permutation-greater-than-target.py
class Solution:
    def lexGreaterPermutation(self, s: str, target: str) -> str:
        cnt = [0] * 26
        for ch in s:
            cnt[ord(ch) - ord('a')] += 1
        for ch in target:
            cnt[ord(ch) - ord('a')] -= 1
        bad = 0
        mask = 0
        for c in range(26):
            if cnt[c] < 0:
                bad += 1
            elif cnt[c]:
                mask |= 1 << c
        for i in range(len(target) - 1, -1, -1):
            cur = ord(target[i]) - ord('a')
            cnt[cur] += 1
            if cnt[cur] == 0:
                bad -= 1
            elif cnt[cur] == 1:
                mask |= 1 << cur
            if bad:
                continue
            up = mask >> (cur + 1)
            if not up:
                continue
            nxt = cur + 1 + (up & -up).bit_length() - 1
            cnt[nxt] -= 1
            ans = list(target[:i])
            ans.append(chr(nxt + ord('a')))
            for c in range(26):
                ans.extend(chr(c + ord('a')) * cnt[c])
            return ''.join(ans)
        return ""

File: 3720-lexicographically-smallest-permutation-greater-than-target/3720-lexicographically-smallest-permutation-greater-than-target.cpp
class Solution {
public:
    string lexGreaterPermutation(string s, string target) {
        int cnt[26] = {};
        int mask = 0, bad = 0;
        for (char ch : s)
            cnt[ch - 'a']++;
        for (char ch : target)
            cnt[ch - 'a']--;
        for (int c = 0; c < 26; c++) {
            if (cnt[c] < 0)
                bad++;
            else if (cnt[c] > 0)
                mask |= 1 << c;
        }
        for (int i = target.size() - 1; i >= 0; i--) {
            int cur = target[i] - 'a';
            cnt[cur]++;
            if (cnt[cur] == 0)
                bad--;
            else if (cnt[cur] == 1)
                mask |= 1 << cur;
            if (bad > 0)
                continue;
            int higher = mask >> (cur + 1);
            if (!higher)
                continue;
            int next = cur + 1 + __builtin_ctz(higher);
            cnt[next]--;
            string ans = target.substr(0, i);
            ans += char('a' + next);
            for (int c = 0; c < 26; c++) {
                ans.append(cnt[c], char('a' + c));
            }
            return ans;
        }
        return "";
    }
};